Source/JavaScriptCore/runtime/MegamorphicCache.h

```diff
@@ -0,0 +1,150 @@
+    WTF_MAKE_NONCOPYABLE(MegamorphicCache);
+public:
+    static constexpr uint32_t primarySize = 2048;
+    static constexpr uint32_t secondarySize = 512;
+    static_assert(hasOneBitSet(primarySize), "size should be a power of two.");
+    static_assert(hasOneBitSet(secondarySize), "size should be a power of two.");
+    static constexpr uint32_t primaryMask = primarySize - 1;
+    static constexpr uint32_t secondaryMask = secondarySize - 1;
+
+    static constexpr uint16_t invalidEpoch = 0;
+    static constexpr PropertyOffset maxOffset = UINT16_MAX;
+
+    struct Entry {
+        static ptrdiff_t offsetOfUid() { return OBJECT_OFFSETOF(Entry, m_uid); }
+        static ptrdiff_t offsetOfStructureID() { return OBJECT_OFFSETOF(Entry, m_structureID); }
+        static ptrdiff_t offsetOfEpoch() { return OBJECT_OFFSETOF(Entry, m_epoch); }
+        static ptrdiff_t offsetOfOffset() { return OBJECT_OFFSETOF(Entry, m_offset); }
+        static ptrdiff_t offsetOfHolder() { return OBJECT_OFFSETOF(Entry, m_holder); }
+
+        void initAsMiss(StructureID structureID, UniquedStringImpl* uid, uint16_t epoch)
+        {
+            m_uid = uid;
+            m_structureID = structureID;
+            m_epoch = epoch;
+            m_offset = 0;
+            m_holder = nullptr;
+        }
+
+        void initAsHit(StructureID structureID, UniquedStringImpl* uid, uint16_t epoch, JSCell* holder, uint8_t offset, bool ownProperty)
+        {
+            m_uid = uid;
+            m_structureID = structureID;
+            m_epoch = epoch;
+            m_offset = offset;
+            m_holder = (ownProperty) ? JSCell::seenMultipleCalleeObjects() : holder;
+        }
+
+        RefPtr<UniquedStringImpl> m_uid;
+        StructureID m_structureID { };
+        uint16_t m_epoch { invalidEpoch };
+        uint16_t m_offset { 0 };
+        JSCell* m_holder { nullptr };
+    };
+
+    static ptrdiff_t offsetOfPrimaryEntries() { return OBJECT_OFFSETOF(MegamorphicCache, m_primaryEntries); }
+    static ptrdiff_t offsetOfSecondaryEntries() { return OBJECT_OFFSETOF(MegamorphicCache, m_secondaryEntries); }
+    static ptrdiff_t offsetOfEpoch() { return OBJECT_OFFSETOF(MegamorphicCache, m_epoch); }
+
+    MegamorphicCache() = default;
+
+#if CPU(ADDRESS64) && !ENABLE(STRUCTURE_ID_WITH_SHIFT)
+    // Because Structure is allocated with 16-byte alignment, we should assume that StructureID's lower 4 bits are zeros.
+    static constexpr unsigned structureIDHashShift1 = 4;
+#else
+    // When using STRUCTURE_ID_WITH_SHIFT, all bits can be different. Thus we do not need to shift the first level.
+    static constexpr unsigned structureIDHashShift1 = 0;
+#endif
+    static constexpr unsigned structureIDHashShift2 = structureIDHashShift1 + 11;
+
+    static constexpr unsigned structureIDHashShift3 = 9 + 4;
+
+    ALWAYS_INLINE static uint32_t primaryHash(StructureID structureID, UniquedStringImpl* uid)
+    {
+        uint32_t sid = bitwise_cast<uint32_t>(structureID);
```

Comment:
It would be interesting to collect a list of StructureIDs and measure how uniform these hash functions actually are

Reply:
Currently, we made it work well for various benchmark cases. And so far, primary / secondary hash can cover 95~% cases.